Content.Server/Botany/Systems/PlantAnalyzerSystem.cs

```diff
@@ -0,0 +1,261 @@
+using Content.Shared.Interaction;
+using Content.Shared.PlantAnalyzer;
+using Robust.Server.GameObjects;
+using Robust.Shared.Audio.Systems;
+using Robust.Shared.Player;
+using Robust.Shared.Prototypes;
+using System.Linq;
+using System.Threading;
+
+namespace Content.Server.Botany.Systems;
+
+public sealed class PlantAnalyzerSystem : EntitySystem
+{
+    [Dependency] private readonly PowerCellSystem _cell = default!;
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly SharedDoAfterSystem _doAfterSystem = default!;
+    [Dependency] private readonly UserInterfaceSystem _uiSystem = default!;
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+    private Dictionary<EntityUid, CancellationTokenWrapper> _cancellationTokenSources = new Dictionary<EntityUid, CancellationTokenWrapper>();
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<PlantAnalyzerComponent, AfterInteractEvent>(OnAfterInteract);
+        SubscribeLocalEvent<PlantAnalyzerComponent, PlantAnalyzerDoAfterEvent>(OnDoAfter);
+        SubscribeLocalEvent<PlantAnalyzerComponent, PlantAnalyzerSetMode>(OnModeSelected);
+    }
+
+    private void OnAfterInteract(Entity<PlantAnalyzerComponent> plantAnalyzer, ref AfterInteractEvent args)
+    {
+        if (args.Target == null || !args.CanReach || !HasComp<SeedComponent>(args.Target) && !HasComp<PlantHolderComponent>(args.Target) || !_cell.HasActivatableCharge(plantAnalyzer, user: args.User))
+            return;
+
+        _audio.PlayPvs(plantAnalyzer.Comp.ScanningBeginSound, plantAnalyzer);
+
+        var cancellationTokenSource = new CancellationTokenSource();
+        var cancellationTokenWrapper = new CancellationTokenWrapper(cancellationTokenSource);
+
+        _cancellationTokenSources[plantAnalyzer] = cancellationTokenWrapper;
+
+        if (plantAnalyzer.Comp.AdvancedScan)
+        {
+            _doAfterSystem.TryStartDoAfter(new DoAfterArgs(EntityManager, args.User, plantAnalyzer.Comp.AdvScanDelay,
+                new PlantAnalyzerDoAfterEvent { CancellationTokenWrapper = cancellationTokenWrapper }, plantAnalyzer, target: args.Target, used: plantAnalyzer)
+            {
+                BreakOnTargetMove = true,
+                BreakOnUserMove = true,
+                NeedHand = true,
+            });
+        }
+        else
+        {
+            _doAfterSystem.TryStartDoAfter(new DoAfterArgs(EntityManager, args.User, plantAnalyzer.Comp.ScanDelay,
+                new PlantAnalyzerDoAfterEvent { CancellationTokenWrapper = cancellationTokenWrapper }, plantAnalyzer, target: args.Target, used: plantAnalyzer)
+            {
+                BreakOnTargetMove = true,
+                BreakOnUserMove = true,
+                NeedHand = true
+            });
+        }
+    }
+
+    private void OnDoAfter(Entity<PlantAnalyzerComponent> component, ref PlantAnalyzerDoAfterEvent args)
```

Comment:
```suggestion
    private void OnDoAfter(Entity<PlantAnalyzerComponent> ent, ref PlantAnalyzerDoAfterEvent args)
```

(do this everywhere)